Cannot call C++ __qpu__ function from within another kernel if parameters contain vector of cudaq::measure_results

### Required prerequisites

- [X] Consult the [security policy](https://github.com/NVIDIA/cuda-quantum/security/policy). If reporting a security vulnerability, do not report the bug using this form. Use the process described in the policy to report the issue.
- [X] Make sure you've read the [documentation](https://nvidia.github.io/cuda-quantum/latest). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] If possible, make a PR with a failing test to give us a starting point to work on!

### Describe the bug

Calling a function with `const std::vector<cudaq::measure_result> &result_vec` parameters seems to not work when running with `--enable-mlir`.

### Steps to reproduce the bug

Consider the following file:
```cpp
#include "cudaq.h"

bool xor_result(const std::vector<cudaq::measure_result> &result_vec) __qpu__ {
  bool result = false;
  for (auto x : result_vec)
    result ^= x;
  return result;
}

bool kernel() __qpu__ {
  cudaq::qvector q(7);
  std::vector<cudaq::measure_result> mz_res = mz(q);
  bool res = xor_result(mz_res);
  return res;
}

int main(int argc, char *argv[]) {
  printf("Result: %d\n", static_cast<int>(kernel()));
  return 0;
}
```
That produces the following compilation error:
```bash
$ nvq++ --enable-mlir issue.cpp
issue.cpp:21:14: error: 'func.call' op kernel call was not inlined, possible recursion in call tree
issue.cpp:21:14: note: see current operation: %3 = "func.call"(%2) {callee = @__nvqpp__mlirgen__function_xor_result._Z10xor_resultRKSt6vectorIbSaIbEE} : (!cc.stdvec<i1>) -> i1
failed: "/workspaces/cuda-quantum/build/bin/cudaq-opt --pass-pipeline=builtin.module(func.func(unwind-lowering),canonicalize,lambda-lifting,func.func(memtoreg{quantum=0}),canonicalize,apply-op-specialization,kernel-execution,aggressive-early-inlining,func.func(quake-add-metadata),device-code-loader{use-quake=1},expand-measurements,func.func(lower-to-cfg),canonicalize,cse) issue.qke -o issue.qke.v8G01q"
```
There is no recursion in the call tree.

Also, interestingly enough, if one were to remove the `__qpu__` attribute on the `xor_result` function, the program compiles but produces a `Segmentation fault`, which is undesirable. We should throw a cleaner error message in that case.

### Expected behavior

The program provide in the above section should compile and run.

### Is this a regression? If it is, put the last known working version (or commit) here.

Not a regression

### Environment

- **CUDA Quantum version**: latest (f7983754b)
- **Python version**: 
- **C++ compiler**: 
- **Operating system**: 


### Suggestions

_No response_

--- a/include/cudaq/Optimizer/Dialect/Common/InlinerInterface.h
+++ b/include/cudaq/Optimizer/Dialect/Common/InlinerInterface.h
@@ -18,21 +18,21 @@ namespace cudaq {
 
 /// Generic opt-in with inlining any operation. Inlining kernels is important to
 /// build up larger functions to translate into quantum circuits.
-/// Do not allow kernel entry points to be inlined into their callers, typically
+/// Do not allow kernel entry points to be inlined into their callers while in
 /// the thunk function.
 struct EnableInlinerInterface : public mlir::DialectInlinerInterface {
   using DialectInlinerInterface::DialectInlinerInterface;
 
-  bool isLegalToInline(mlir::Region *, mlir::Region *src, bool,
+  bool isLegalToInline(mlir::Region *dest, mlir::Region *src, bool,
                        mlir::IRMapping &) const final {
-    if (auto func = src->getParentOfType<mlir::func::FuncOp>())
-      return !(func->hasAttr(cudaq::entryPointAttrName));
+    if (auto destFunc = dest->getParentOfType<mlir::func::FuncOp>())
+      if (destFunc.getName().ends_with(".thunk"))
+        if (auto srcFunc = src->getParentOfType<mlir::func::FuncOp>())
+          return !(srcFunc->hasAttr(cudaq::entryPointAttrName));
     return true;
   }
   bool isLegalToInline(mlir::Operation *op, mlir::Region *, bool,
                        mlir::IRMapping &) const final {
-    if (auto func = op->getParentOfType<mlir::func::FuncOp>())
-      return !(func->hasAttr(cudaq::entryPointAttrName));
     return true;
   }
   bool isLegalToInline(mlir::Operation *call, mlir::Operation *callable,

--- a/include/cudaq/Optimizer/Dialect/Common/InlinerInterface.h
+++ b/include/cudaq/Optimizer/Dialect/Common/InlinerInterface.h
@@ -31,8 +31,12 @@ struct EnableInlinerInterface : public mlir::DialectInlinerInterface {
           return !(srcFunc->hasAttr(cudaq::entryPointAttrName));
     return true;
   }
-  bool isLegalToInline(mlir::Operation *op, mlir::Region *, bool,
+  bool isLegalToInline(mlir::Operation *op, mlir::Region *dest, bool,
                        mlir::IRMapping &) const final {
+    if (auto destFunc = dest->getParentOfType<mlir::func::FuncOp>())
+      if (destFunc.getName().ends_with(".thunk"))
+        if (auto srcFunc = op->getParentOfType<mlir::func::FuncOp>())
+          return !(srcFunc->hasAttr(cudaq::entryPointAttrName));
     return true;
   }
   bool isLegalToInline(mlir::Operation *call, mlir::Operation *callable,

--- a/lib/Optimizer/Dialect/Quake/QuakeOps.cpp
+++ b/lib/Optimizer/Dialect/Quake/QuakeOps.cpp
@@ -914,7 +914,7 @@ bool cudaq::EnableInlinerInterface::isLegalToInline(Operation *call,
   if (auto applyOp = dyn_cast<quake::ApplyOp>(call))
     if (applyOp.applyToVariant())
       return false;
-  return !(callable->hasAttr(cudaq::entryPointAttrName));
+  return true;
 }
 
 using EffectsVectorImpl =

--- a/lib/Optimizer/Dialect/Quake/QuakeOps.cpp
+++ b/lib/Optimizer/Dialect/Quake/QuakeOps.cpp
@@ -914,6 +914,10 @@ bool cudaq::EnableInlinerInterface::isLegalToInline(Operation *call,
   if (auto applyOp = dyn_cast<quake::ApplyOp>(call))
     if (applyOp.applyToVariant())
       return false;
+  if (auto destFunc = call->getParentOfType<mlir::func::FuncOp>())
+    if (destFunc.getName().ends_with(".thunk"))
+      if (auto srcFunc = call->getParentOfType<mlir::func::FuncOp>())
+        return !(srcFunc->hasAttr(cudaq::entryPointAttrName));
   return true;
 }
 

--- a/targettests/Kernel/inline-qpu-func.cpp
+++ b/targettests/Kernel/inline-qpu-func.cpp
@@ -1,0 +1,36 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+// RUN: nvq++ %cpp_std --enable-mlir %s -o %t && %t | FileCheck %s
+
+#include "cudaq.h"
+
+// This function has no cudaq::qubit's in the parameter list, so it will be
+// tagged as a possible cudaq-entrypoint kernel. Make sure we can still inline
+// it if called from another kernel.
+bool xor_result(const std::vector<cudaq::measure_result> &result_vec) __qpu__ {
+  bool result = false;
+  for (auto x : result_vec)
+    result ^= x;
+  return result;
+}
+
+bool kernel() __qpu__ {
+  cudaq::qvector q(7);
+  x(q);
+  std::vector<cudaq::measure_result> mz_res = mz(q);
+  bool res = xor_result(mz_res);
+  return res;
+}
+
+int main(int argc, char *argv[]) {
+  printf("Result: %d\n", static_cast<int>(kernel()));
+  return 0;
+}
+
+// CHECK: Result: 1

